### ISSUE TITLE
Human readable json format for the patches

### ISSUE THREAD

**Issue by Haerezis** [AEFeinstein/GathererScraper#4]
Hello,

I was wondering if there were a reason behind the current json format for the patches ?

I want to start working on adding multilingual support to the patches so that I can start doing the same for MTG Familiar. But I was puzzled when I saw the current json format.

If I do a commit both on GathererScraper and MTG Familiar to change the json format to make it human readable, would you accept it?

**Comment by AEFeinstein** [comment]
The json format was a poor attempt at keeping patches as small as possible. Was it necessary with gzip? Probably not, but young-me had an embedded-centric mentality and made some questionable decisions.

The only potential pitfall of changing the patch format would be that prior builds of Familiar (2.x, pre-ICS) wouldn't be able to receive new patches (I'm not maintaining two sets of patches). Looking at the install numbers, there are few enough of those users that the only ones to really be impacted would be users who don't regularly update apps. I can handle their emails about why Familiar isn't updating anymore.

Speaking of questionable decisions, the json parser in Familiar is trash. There's no reason for stepping through a JsonReader when Gson can populate and return Java objects all by itself. The JsonReader is actually fragile, and I've been meaning to replace it for some time. I guess now is the time.

So in short, yes, if you submit pull requests for both Scraper and Familiar (and use [fromJson()](https://google.github.io/gson/apidocs/com/google/gson/Gson.html#fromJson-java.io.Reader-java.lang.Class-)), I'd accept it.

**Comment by AEFeinstein** [comment]
Link to Familiar ticket: https://github.com/AEFeinstein/mtg-familiar/issues/14#issuecomment-276531984

**Comment by Haerezis** [comment]
Thanks for your answer, I'll start to work on the json format but don't expect it to be that quick, I just started a new job and it's possible I won't have time to make a lot of progress in that area. But at least I know that I can start working on it and it won't be a waste.

**Comment by AEFeinstein** [comment]
If you don't have much time, you can just do the GathererScraper part. Looking at the code, if patches become human readable, there's redundancy between the JsonCard/Card and JsonExpansion/Expansion classes. If you get rid of the Json* classes, I should be able to use the new class files in Familiar quite easily.

**Comment by Haerezis** [comment]
I've started to work on the patches-v2 format in mtg-familiar.
I think the code should work but I've not tested it yet, and I can't find a way to test the import of a new set. Is there a way to do it without fetching the gzip from the url stored in patches.json ?

**Comment by AEFeinstein** [comment]
I think the easiest thing to do would be to change the patch URLs in GathererScraper to
`entry.mURL = "https://raw.githubusercontent.com/AEFeinstein/GathererScraper/master/patches-v2/" + exp.mCode_gatherer + ".json.gzip";`
Or your branch instead of AEFeinstein. Then have Familiar check these files instead of the Google Sites ones (or the ones in your branch):
https://raw.githubusercontent.com/AEFeinstein/GathererScraper/master/patches.json
https://raw.githubusercontent.com/AEFeinstein/GathererScraper/master/legality.json

**Comment by AEFeinstein** [comment]
I did also change the json format for legality.json, so don't be surprised if that fails. Familiar shouldn't have to check any of the other json files that it previously did.

**Comment by AEFeinstein** [comment]
Are you working on mtg-familiar in a fork?

**Comment by Haerezis** [comment]
Yes I'm currently working on it. I had some bugs that were completely mystery for me, but it was due to the change of json format of the manifest file. I'm ashamed to admit that it took me a while to notice the difference. I'm kind of busy up until friday evening but I'll be free this weekend so I think I'll work on it to finish a pull request this weekend.

I think you can expect a pull request from me around the end of the weekend.
If that's not fast enough for you, let me know.

**Comment by Haerezis** [comment]
Sorry again, with my new job, I often come home later than usual and a little bit tired, so I don't make much progress during the week. And last weekend I was exceptionally busy with some Magic tournaments :/ But you have to know that I'm really motivated to work on that and later on multi-lingual support (because frankly, I really need that feature sometime).

**Comment by AEFeinstein** [comment]
That's fine, there's no rush. I just didn't see a fork under your GitHub
account so I wasn't sure if you were working on it, or if I should. Good
luck with the new job!

**Comment by Haerezis** [comment]
Ok so I've pushed a commit for mtg familiar, but we need to discuss a few things before thinking about merging it.

With my commit, I've discovered that the Conspiracy set (CNS) doesn't have a correct date in its patch (we have 6/2014 but we should have 06/2014).
Where I tried to re-scrap to regenerate it, I discovered that you cannot scrape CNS anymore : when you scrap, you use the gatherer name, but the gatherer name of CNS has a em dash (long dash), and not a normal dash in its name, and when you get the list of set from gatherer, you replace the em dash with a normal dash.
The problem is : when you scrap the card for the set, you scrap using the modifed name (without em dash), and gatherer return 0 card for this name, but return the corrects amount of card for the version with an em dash.

If everything was perfect, we could simply use the em dash version in both the scraper AND mtg-familiar. Unfortunatly, it seems like em dash is not correctly displayed in mtg-familiar, and it push CNS at the top of the set list in the search screen (dunno why).

All in all, I think the best solution would be to use the em dash version in the scraper, but when we generate the patches.json and the individual patches, with replace it with a normal dash.

**Comment by AEFeinstein** [comment]
Okay, I fixed the Conspiracy dash issue by only replacing the non-ASCII dash when the program writes the patch file, not when it reads the set lists. I updated CNS.json.gzip in master. That was my mistake.

I also noticed that the patches used to use Unix timestamps for the date, but now it uses a human readable String. That's fine by me, but the parsing shouldn't be sensitive to leading zeros. Something like this would work:

```
int month = Integer.parseInt(dateStr.split("/")[0]);
int year  = Integer.parseInt(dateStr.split("/")[1]);
```

**Comment by Haerezis** [comment]
Ok so I was half wrong :s sorry.
I was a little lost between the expansion name in a card, in patches.json file, etc... I was changing the set name (to use the em dash), but not the digest, so mtg-familiar was behaving weirdly..

SO. The good news is : you just have to NOT replace the em dash when getting the set list from gatherer. I've just tested : it will generate a correct expansion.json, patches.json and the CNS.json.gzip.

When mtg-gatherer fetch the patches.json, it will correctly get the the set name, and its patch file, and there is not problem in the set name display. I've tested it with my mtg-familiar commit using the patch-v2 format (and fetching to my raw github).

You can look at the last commit I've made on GathererScraper.

And about the date, in my opinion, a date without a leading zero in case of month < october is not formatted correctly. The month number NEED to be composed of 2 integers. Maybe instead of a String of the date, we should use a Date (and gson will take care of the serialization/deserialization into string in the json). I can take care of that if you want.
I understand why you used timestamp (because of sqlite on android), but you can get them from Date when storing the card in mtg-familiar.

What do you think?

**Comment by AEFeinstein** [comment]
I think that the patch should still use a Unix timestamp. They're easy to store in a database, compare, and serialize. The only downside is that they're hard to look at. The date column in GathererScraper can do a conversion between timestamp and human readable date.

**Comment by Haerezis** [comment]
Ok. I won't argue with the project owner. But in that case, we should use seconds and not milliseconds for the timestamp. The unix timstamp is historically in seconds, not in milliseconds, and anyone reading a timestamp will think so (at least it was my first thought at the beginning).

We also need to rename the field from 'date' to 'timestamp' so that it's crystal clear to users that the film is a unix timestamp, and not a weird custom date format.

**Comment by AEFeinstein** [comment]
I agree with both replacing milliseconds with seconds and renaming date to timestamp.

**Comment by Haerezis** [comment]
Ok. i'll take care of that. And I'll also remove the replacement of the em dash with a dish in the scraper, and also update mtg-familiar. You can expect a PR pour the scraper and mtg-familiar tomorrow.

**Comment by AEFeinstein** [comment]
si tu veux parler en francais, je sais un peu :)

**Comment by Haerezis** [comment]
Haha, est ce que tu proposes parce que tu trouves que mon anglais est horrible? x)

I have a question : where and how do you generate the digests.json, tcgnames.json and canBeFoil.json files ? Shouldn't they be generated by the scraper? Because I have to comment the code checking the digest from digests.json (yours because I don't know how to generate one) with the digest in the db of the app, because for me the digest in digests.json is wrong.

**Comment by AEFeinstein** [comment]
Parce-que tu utilises le mot "pour" en anglais! Mais ton anglais est meilleur que mon francais.

The data from digests.json has been added to patches.json, and the data from tcgnames.json and canBeFoil.json have been added to the individual patch files. Those three files have been removed intentionally. The old json parser in MTG Familiar would crash if data fields were added, removed, or moved, so when I wanted to add digests, TCGPlayer.com set names, and the foil information, I had to do it with new files. It wasn't elegant. Thankfully gson is more tolerant.

**Comment by Haerezis** [comment]
ha, sorry for the "pour", I never noticed.

Ok for the differents files. So I'll also remove their use from the code of mtg-familiar, and directly use the content of patches.json.

**Comment by Haerezis** [comment]
I've proposed 2 PR  for the json format update. One for mtg-familiar and one last for the GathererScraper.

**Comment by AEFeinstein** [comment]
This all looks good. I'll close this ticket and open another for the foreign names
